src/main/java/io/pinecone/PineconeIndexOperationClient.java

```diff
@@ -1,52 +1,88 @@
+import io.pinecone.exceptions.PineconeConfigurationException;
 import io.pinecone.model.CreateIndexRequest;
-import org.asynchttpclient.AsyncHttpClient;
-import org.asynchttpclient.DefaultAsyncHttpClient;
+import okhttp3.*;
 
 import java.io.IOException;
 
 public class PineconeIndexOperationClient {
-    private AsyncHttpClient client;
-    private PineconeClientConfig clientConfig;
-    private String url;
+    private final OkHttpClient client;
+    private final PineconeClientConfig clientConfig;
+    private final String url;
+    public static final String ACCEPT_HEADER = "accept";
+    public static final String API_KEY_HEADER_NAME = "Api-Key";
+    public static final String BASE_URL_PREFIX = "https://controller.";
+    public static final String BASE_URL_SUFFIX = ".pinecone.io/databases/";
+    public static final String CONTENT_TYPE = "content-type";
+    public static final String CONTENT_TYPE_JSON = "application/json";
+    public static final String TEXT_PLAIN = "text/plain";
 
-    PineconeIndexOperationClient(PineconeClientConfig clientConfig, AsyncHttpClient client) {
-        this.clientConfig = clientConfig;
+    private PineconeIndexOperationClient(PineconeClientConfig clientConfig, OkHttpClient client, String url) {
         this.client = client;
-        this.url = "https://controller." + clientConfig.getEnvironment() + ".pinecone.io/databases/";
+        this.clientConfig = clientConfig;
+        this.url = url;
+    }
+
+    public PineconeIndexOperationClient(PineconeClientConfig clientConfig, OkHttpClient client) {
+        this(clientConfig, client, createUrl(clientConfig));
     }
 
     public PineconeIndexOperationClient(PineconeClientConfig clientConfig) {
-        this.clientConfig = clientConfig;
-        client = new DefaultAsyncHttpClient();
-        this.url = "https://controller." + clientConfig.getEnvironment() + ".pinecone.io/databases/";
+        this(clientConfig, new OkHttpClient());
+    }
+
+    private static String createUrl(PineconeClientConfig clientConfig) {
+        if (clientConfig.getApiKey() == null || clientConfig.getEnvironment() == null) {
+            throw new PineconeConfigurationException("Both API key and environment name are required for index operations.");
+        }
+
+        return BASE_URL_PREFIX + clientConfig.getEnvironment() + BASE_URL_SUFFIX;
     }
 
     public void deleteIndex(String indexName) throws IOException {
-        System.out.println("Sending delete index request:");
-        client.prepare("DELETE", url + indexName)
-                .setHeader("accept", "text/plain")
-                .setHeader("Api-Key", clientConfig.getApiKey())
-                .execute()
-                .toCompletableFuture()
-                .join();
+        Request request = new Request.Builder()
+                .url(url + indexName)
+                .delete()
+                .addHeader(ACCEPT_HEADER, TEXT_PLAIN)
+                .addHeader(API_KEY_HEADER_NAME, clientConfig.getApiKey())
```

Comment:
It seems like a bummer that you end up writing a lot of similar boilerplate for each request. What do you think about making a helper method that encapsulates some of these steps that happen in every request we make?

Something like this (??? is a placeholder for the return type, which I can't know without some help from the docs or IDE)

```
private ??? initRequest(String path) {
  return new Request.Builder()
         .url(url + path)
         .addHeader(ACCEPT_HEADER, TEXT_PLAIN)
         .addHeader(API_KEY_HEADER_NAME, clientConfig.getApiKey())
}
```

and then usage would be something like

```
Request request = initRequest(indexName)
                                   .delete()
                                   .build()
```

Reply:
I agree with you completely, so I've analyzed all index and collection api calls, and encapsulated the request builder method to support all data plane operations.